File: spring-data-requery-java/src/test/java/com/coupang/springframework/data/requery/domain/upsert/AbstractUpsertAddress.java
package com.coupang.springframework.data.requery.domain.upsert;

import com.coupang.kotlinx.objectx.ToStringBuilder;
import com.coupang.springframework.data.requery.domain.AbstractComponent;
import io.requery.Embedded;
import io.requery.Transient;
import lombok.Getter;
import org.jetbrains.annotations.NotNull;

/**
 * @author Diego on 2018. 6. 10..
 */
@Getter
@Embedded
public abstract class AbstractUpsertAddress extends AbstractComponent {

    protected String address;
    protected String city;
    protected String zipcode;

    @Override
    public int hashCode() {
        return zipcode.hashCode();
    }

    @NotNull
    @Override
    @Transient
    protected ToStringBuilder buildStringHelper() {
        return super.buildStringHelper()
            .add("zipcode", zipcode)
            .add("city", city)
            .add("address", address);
    }
}

File: src/test/kotlin/com/coupang/springframework/data/requery/domain/upsert/AbstractUpsertAddress.kt
package com.coupang.springframework.data.requery.domain.upsert

import com.coupang.kotlinx.objectx.ToStringBuilder
import com.coupang.springframework.data.requery.domain.AbstractComponent
import io.requery.Column
import io.requery.Embedded
import io.requery.Transient

/**
 * AbstractUpsertAddress
 *
 * @author debop@coupang.com
 * @since 18. 5. 31
 */
@Embedded
abstract class AbstractUpsertAddress: AbstractComponent() {

    @get:Column
    abstract var address: String?

    @get:Column
    abstract var city: String?

    @get:Column(nullable = false)
    abstract var zipcode: String

    override fun hashCode(): Int {
        return zipcode.hashCode()
    }

    @Transient
    override fun buildStringHelper(): ToStringBuilder {
        return super.buildStringHelper()
            .add("zipcode", zipcode)
            .add("city", city)
            .add("address", address)
    }
}